python/src/nanoarrow/visitor.py

```diff
@@ -15,68 +15,186 @@
+    >>> na.Array([1, None, 3], na.int32()).to_column(na.nulls_forbid())
+    Traceback (most recent call last):
+    ...
+    ValueError: Null present with null_handler=nulls_forbid()
     """
-    return ListBuilder.visit(obj, schema)
 
+    def handle(is_valid, data):
+        if len(is_valid) > 0:
+            raise ValueError("Null present with null_handler=nulls_forbid()")
+
+        return data
+
+    return handle
 
-def to_columns(obj, schema=None) -> Tuple[List[str], List[Sequence]]:
-    """Convert ``obj`` to a ``list()` of sequences
 
-    Converts a stream of struct arrays into its column-wise representation
-    such that each column is either a contiguous buffer or a ``list()``.
+def nulls_as_sentinel(sentinel=None):
+    """Sentinel null handler
 
-    Paramters
-    ---------
-    obj : array stream-like
-        An array-like or array stream-like object as sanitized by
-        :func:`c_array_stream`.
-    schema : schema-like, optional
-        An optional schema, passed to :func:`c_array_stream`.
+    A null handler that assigns a sentinel to null values. This is
+    done using numpy using the expression ``data[~is_valid] = sentinel``.
+    The default sentinel value will result in ``nan`` assigned to null
+    values in numeric and boolean outputs.
+
+    Parameters
+    ----------
+    sentinel : scalar, optional
+        The value with which nulls should be replaced.
 
     Examples
     --------
 
     >>> import nanoarrow as na
-    >>> from nanoarrow import visitor
-    >>> import pyarrow as pa
-    >>> array = pa.record_batch([pa.array([1, 2, 3])], names=["col1"])
-    >>> names, columns = visitor.to_columns(array)
-    >>> names
-    ['col1']
-    >>> columns
-    [[1, 2, 3]]
+    >>> na.Array([1, 2, 3], na.int32()).to_column(na.nulls_as_sentinel())
+    array([1, 2, 3], dtype=int32)
+    >>> na.Array([1, None, 3], na.int32()).to_column(na.nulls_as_sentinel())
+    array([ 1., nan,  3.])
+    >>> na.Array([1, None, 3], na.int32()).to_column(na.nulls_as_sentinel(-999))
+    array([   1, -999,    3], dtype=int32)
     """
-    return ColumnsBuilder.visit(obj, schema)
+    import numpy as np
+
+    def handle(is_valid, data):
+        is_valid = np.array(is_valid, copy=False)
+        data = np.array(data, copy=False)
+
+        if len(is_valid) > 0:
+            out_type = np.result_type(data, sentinel)
+            data = np.array(data, dtype=out_type, copy=True)
+            data[~is_valid] = sentinel
+            return data
+        else:
+            return data
+
+    return handle
+
+
+def nulls_separate() -> Callable[[CBuffer, Sequence], Tuple[CBuffer, Sequence]]:
+    """Return nulls as a tuple of is_valid, data
+
+    A null handler that returns its components.
+
+    Examples
+    --------
+
+    >>> import nanoarrow as na
+    >>> na.Array([1, 2, 3], na.int32()).to_column(na.nulls_separate())
+    (nanoarrow.c_lib.CBuffer(uint8[0 b] ), nanoarrow.c_lib.CBuffer(int32[12 b] 1 2 3))
```

Comment:
Should we return `None` here instead of an empty buffer? Because if you convert both outputs to numpy arrays, you can have numpy arrays of different length (of course, it's not that one of checking for None or for empty / len 0 is much harder than the other)